pkg/util/async/bounded_frequency_runner.go

```diff
@@ -223,17 +291,16 @@ func (bfr *BoundedFrequencyRunner) tryRun() {
 	}
 
 	// It can't run right now, figure out when it can run next.
-
 	elapsed := bfr.timer.Since(bfr.lastRun)    // how long since last run
 	nextPossible := bfr.minInterval - elapsed  // time to next possible run
-	nextScheduled := bfr.maxInterval - elapsed // time to next periodic run
+	nextScheduled := bfr.timer.Remaining()     // time to next scheduled run
 	klog.V(4).Infof("%s: %v since last run, possible in %v, scheduled in %v", bfr.name, elapsed, nextPossible, nextScheduled)
 
+	// It's hard to avoid race conditions in the unit tests unless we always reset
+	// the timer here, even when it's unchanged
 	if nextPossible < nextScheduled {
```

Comment:
(Note that the comment applies to the old code as well; the `if` statement here used to always be true, since `nextPossible < nextScheduled` → `bfr.minInterval - elapsed < bfr.maxInterval - elapsed` → `bfr.minInterval < bfr.maxInterval`, which is enforced by `NewBoundedFrequencyRunner`. So we were always resetting the timer before too, it just wasn't immediately obvious.)